Experimental - Add content attribute controls to "Content" sidebar in contentOnly

getdave opened the pull request:
This PR has been co-authored by @jeryj.


## What

This PR _explores_ (this is a WIP/prototype!) an **automatic approach** to displaying standardized UI controls for attributes with `role: 'content'` in the block quick navigation sidebar. Unlike the alternative approach in [#71577](https://github.com/WordPress/gutenberg/pull/71577), this implementation requires **no changes to existing block.json files** and works automatically with all blocks (core and third-party).

@jeryj and I are not sure which approach is best so we felt that prototyping this might help contributors to decide on which approach is best.

Closes [#71557](https://github.com/WordPress/gutenberg/issues/71557): Add automatic content attribute controls to block quick navigation sidebar

### Key Features

- **Automatic Control Inference**: Automatically detects and renders appropriate UI controls based on attribute type and naming patterns
- **DataForm Integration**: Uses `@wordpress/dataviews` DataForm components as suggested in [#71577#issuecomment-3274658200](https://github.com/WordPress/gutenberg/pull/71577#issuecomment-3274658200) - cc @youknowriad 
- **Smart "Drill Down" Navigation**: Single-attribute blocks show inline controls; multi-attribute blocks use Navigator sub-views with drill-down functionality
- **Comprehensive Type Support**: Handles `string`, `boolean`, `rich-text`, `number`, and URL attributes with intelligent field type mapping
- **Zero Configuration**: Works immediately with all existing blocks without requiring block authors to update their `block.json`

## Why

The current contentOnly mode lacks standardized controls for editing block content attributes in the sidebar. This creates an inconsistent user experience where users must switch between the canvas and sidebar to edit different aspects of their content.

**This automatic approach addresses the core concern raised in [#71577](https://github.com/WordPress/gutenberg/pull/71577#issuecomment-3274658200)**: ensuring write mode works for third-party blocks without requiring updates from block authors.

### Comparison with Alternative Approach

| Aspect | This PR (Automatic) | [#71577](https://github.com/WordPress/gutenberg/pull/71577) (Manual) |
|--------|-------------------|---------------------------------------------------------------------|
| **Configuration** | Zero - works automatically | Requires `control` property in `block.json` |
| **Third-party compatibility** | ✅ Works immediately | ❌ Requires block updates |
| **Maintenance** | Low - centralized logic | High - per-block configuration |
| **Customization** | Limited to inferred types | Full control per attribute |
| **Adoption barrier** | None | High - requires ecosystem updates |

**We acknowledge that both approaches have merit and the optimal solution may be a hybrid approach** that combines automatic inference with optional manual overrides.

## How

### Automatic Control Inference

The system automatically maps attribute types to appropriate DataForm field types:

- **String attributes** → `text` field
- **Boolean attributes** → `toggle` field  
- **Rich-text attributes** → `text` field (with future RichText support)
- **Number attributes** → `text` field
- **URL attributes** → `url` field (detected by name patterns: `url`, `src`, `href`, `link`)

### Smart UI Layout

- **Single Content Attribute**: Displays inline with block icon and custom label
- **Multiple Content Attributes**: Shows navigation button with chevron arrow, leading to dedicated sub-view
- **Sub-view Navigation**: Uses Navigator component with back button for seamless drill-down experience

### DataForm Integration

Leverages `@wordpress/dataviews` DataForm components for:
- Consistent styling and behavior
- Built-in validation and error handling
- Accessibility compliance
- Future extensibility

### Implementation Details

- **Location**: `packages/block-editor/src/components/block-quick-navigation/index.js`
- **Dependencies**: Added `@wordpress/dataviews` to block-editor package
- **Performance**: Uses `useMemo` for attribute processing and `useSelect` for efficient data access
- **State Management**: Integrates with existing block editor store for attribute updates

## Testing

### Manual Testing

1. **Enable contentOnly mode** in Gutenberg settings
2. **Create a post** with various block types (heading, paragraph, image, etc.)
3. **Select a pattern** containing multiple blocks
4. **Open block quick navigation** sidebar
5. **Verify controls appear** for blocks with `role: 'content'` attributes
6. **Test single-attribute blocks** show inline controls with block icons
7. **Test multi-attribute blocks** show navigation buttons with chevrons
8. **Test sub-view navigation** by clicking multi-attribute blocks
9. **Verify attribute updates** are reflected in the canvas
10. **Test with third-party blocks** to ensure automatic compatibility

### Keyboard Testing

1. **Tab navigation** through all controls
2. **Enter/Space** to activate buttons and toggles
3. **Arrow keys** for navigation between fields
4. **Escape** to close popovers and return to main view
5. **Tab** to navigate between single-attribute controls
6. **Enter** to activate multi-attribute navigation buttons

## Screenshots

<img width="677" height="590" alt="Screenshot 2025-09-10 at 16 52 45" src="https://github.com/user-attachments/assets/508e78b0-e66f-4d65-b8a2-1962c46bf328" />


https://github.com/user-attachments/assets/9cc69001-7999-4f03-9883-1cf63ca69a20


## Breaking Changes

None - this is purely additive functionality.

## Uncertain Elements

- **RichText Integration**: Currently renders as text field; full RichText support may require additional work
- **Custom Control Support**: Future enhancement could allow block authors to override inferred controls
- **Performance**: Large numbers of attributes may impact rendering; monitoring needed
- **Hybrid Approach**: Whether to combine with manual configuration from [#71577](https://github.com/WordPress/gutenberg/pull/71577)

## Related Issues

- Addresses [#71557](https://github.com/WordPress/gutenberg/issues/71557) - Patterns: standardize representation of controls
- Alternative approach: [#71577](https://github.com/WordPress/gutenberg/pull/71577) - Manual control configuration
- Related: [#58233](https://github.com/WordPress/gutenberg/issues/58233) - Features with limited control over block UI
- Similar: [#60779](https://github.com/WordPress/gutenberg/issues/60779) - Block controls standardization

getdave committed:
Add content attribute controls to block quick navigation

- Display standardized UI controls for attributes with role: 'content'
- Support string, boolean, and rich-text attribute types
- String and rich-text render as TextControl
- Boolean renders as ToggleControl
- Controls appear under blockTitle in the sidebar
- Maintains block selection while editing attributes

getdave commented:
Closing in favour of https://github.com/WordPress/gutenberg/pull/71577

getdave committed:
Add Navigator-based sub-view for blocks with multiple content attributes

- Implement Navigator component with home screen and block-specific sub-screens
- Add smart attribute display logic: inline for single attributes, navigation for multiple
- Show chevron arrow for blocks with >1 content attribute to indicate drill-down capability
- Create BlockContentAttributesView for detailed attribute editing in sub-view
- Add SingleContentAttributeControl component for reusable single attribute handling
- Maintain LinkControl with popover functionality for URL attributes
- Ensure only one block sub-view can be open at a time
- Provide 'Back' button to return to main view from sub-view
- Preserve existing inline display behavior for single content attributes

getdave committed:
Tweak title

getdave committed:
Use alternative hook to get corect blockTitlte

getdave committed:
Tweaked display to align stuff

getdave committed:
Remove isPressed state so it does not lock a dark background when returning from deeper level

getdave committed:
Initial effort at using DataForm as suggested in another PR by @youknowriad

jeryj committed:
Use attribute name for label input titles

getdave committed:
Correct field labels and layout

getdave committed:
Align back button

jeryj committed:
Remove role: content from some button and image attributes that should not be write mode editable

I think these attributes were added early in on write mode (see PR# 47936) and incorrectly assigned role content to id, ref, and title attributes.

ramonjd commented on packages/block-library/src/image/block.json:
```diff
@@ -74,8 +73,7 @@
 		},
 		"id": {
-			"type": "number",
-			"role": "content"
```
Do you know why `id` had the content attribute in the first place? 🤔

jeryj replied:
I checked on the source of when it was added in https://github.com/WordPress/gutenberg/pull/47936, and my best guess is that it was so early in the process that it was added because it was so unclear how to make something considered content. From where we are now, it feels safe to say that "id" should not be considered content.

talldan replied:
I think this is the attachment id, and when replacing an image from the media library this will need to be updated (even in content only mode). We'll need a control in the inspector that can change this value. So I'd argue it could be considered as content. I think that's probably why the role was added.

In terms of inferring a control, we probably need to be able to detect the attachment `id` and the `img`/`video` element `src` attributes, and when present show a control that can update these values together. We might also want to match the logic around how the caption and alt text can be set from the attachment's caption.

jeryj replied:
We were discussing identifying the media src and showing a media selection button. We haven't gotten there yet.

It feels like a bug to need id labeled as content. Hopefully using a media selector will set the fields we need.

scruffian replied:
> We'll need a control in the inspector that can change this value. So I'd argue it could be considered as content. I think that's probably why the role was added.

Do we need the attribute to be exposed in the inspector? Could we just update it when the media item itself changes?

ramonjd commented on packages/block-editor/src/components/block-quick-navigation/index.js:
```diff
@@ -19,63 +24,290 @@ import BlockIcon from '../block-icon';
+		if ( blockType?.attributes ) {
+			Object.entries( blockType.attributes ).forEach(
+				( [ attrName, attrDef ] ) => {
+					if ( attrDef.role === 'content' ) {
+						attributes.push( {
+							name: attrName,
+							definition: attrDef,
+							value: block.attributes?.[ attrName ],
+						} );
```
Very clever!

Just checking my skim-reading assumptions: any block attribute could opt-in or opt-out of having content-only editable fields by:

1. Using attributes[ someAttr ].role = 'content'
2. The attribute having a supported type.

jeryj replied:
Yup! That's the idea.

jeryj committed:
WIP: Editor inception - content only editor for sidebar

This is not working. When encountering a paragraph or list block, it creates a content group and passes these blocks to a standalone unstyled editor. The issue is this editor instance is tied to the main editor.

jeryj committed:
Add media selector for images in content attributes

This can keep getting refactored further. The media selector is working though by inferring a selector of img and attribute of src. This can then be expanded for better UI when that is determined.

jasmussen commented:
I'm coming back to this a bit late, I had some AFK. Thanks for the ping. Quick GIF:

![test](https://github.com/user-attachments/assets/824ab0c8-41c9-4638-9dee-ff2c61142fcd)

I guess this is some work that precedes https://github.com/WordPress/gutenberg/pull/71730 yes? I also just reviewed that. Both work well, though I tend to think 71730 works slightly better. The drilldown in the content pattern is impressive, I think this pattern can work especially well with _complex blocks in contentOnly mode_, a la Query Loop or Navigation (think site links), but less so for patterns where arguably it is beneficial to lock things down to what the pattern author has done, and be a bit more limiting in the options. 

The main goal is to provide a simplified editing interface for complex blocks or groups of blocks. For mostly static visual patterns, my instinct would go towards not needing the drilldown.

scruffian commented:
Closing in favour of https://github.com/WordPress/gutenberg/pull/71730
